sycl/doc/extensions/KernelProperties/KernelProperties.asciidoc

```diff
@@ -0,0 +1,525 @@
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
+
+|===
+|Property|Description
+
+|`work_group_size`
+|The `work_group_size` property adds the requirement that the kernel must be
+ launched with the specified work-group size. The number of template arguments
+ in the `Dims` parameter pack must match the dimensionality of the work-group
+ used to invoke the kernel. The order of the template arguments matches the
+ constructor of the `range` class. An implementation may throw an exception
+ for certain combinations of property values, devices and launch configurations,
+ as described for the `reqd_work_group_size` attribute in Table 180 of the
+ SYCL 2020 specification.
+
+|`work_group_size_hint`
+|The `work_group_size_hint` property hints to the compiler that the kernel is
+ likely to be launched with the specified work-group size. The number of
+ template arguments in the `Dims` parameter pack must match the dimensionality
+ of the work-group used to invoke the kernel. The order of the template
+ arguments matches the constructor of the `range` class.
+
+|`sub_group_size`
+|The `sub_group_size` property adds the requirement that the kernel must be
+ compiled and executed with the specified sub-group size. An implementation may
+ throw an exception for certain combinations of property values, devices and
+ launch configurations, as described for the `reqd_sub_group_size` attribute
+ in Table 180 of the SYCL 2020 specification.
+
+|`device_has`
+|The `device_has` property adds the requirement that the kernel must be
+ launched on a device that has all of the aspects listed in the `Aspects`
+ parameter pack. An implementation may throw an exception or issue a
+ diagnostic for certain combinations of aspects, devices and kernel functions,
+ as described for the `requires` attribute in Table 180 of the SYCL 2020
+ specification.
+
+|===
+
+SYCL implementations may introduce additional kernel properties. If any
+combinations of kernel attributes are invalid, this must be clearly documented
+as part of the new kernel property definition.
+
+== Adding a `property_list` to a Kernel Launch
+
+To enable properties to be associated with kernels, this extension adds
+new overloads to each of the variants of `single_task`, `parallel_for` and
+`parallel_for_work_group` defined in the `queue` and `handler` classes. These
+new overloads accept a `sycl::ext::oneapi::property_list` argument. For
+variants accepting a parameter pack, the `sycl::ext::oneapi::property_list`
+argument is inserted immediately prior to the parameter pack; for variants not
+accepting a parameter pack, the `sycl::ext::oneapi::property_list` argument is
+inserted immediately prior to the kernel function.
+
+The overloads introduced by this extension are listed below:
+```c++
+namespace sycl {
+class queue {
+ public:
+  template <typename KernelName, typename KernelType, typename PropertyList>
+  event single_task(PropertyList properties, const KernelType &kernelFunc);
```

Comment:
I think we're following the guidelines in the paragraph above that one:

> In some cases, an extension does not have the freedom to choose a specific function name. For example, this could happen if the extension adds a new constructor overload for an existing SYCL class. In cases like this, the extension should ensure that one of the function parameters has a type that is defined in the extension’s namespace. For example, the Acme vendor could add a new constructor for `sycl::context` with the signature `context(ext::acme::frobber &)`.

We're adding an argument to `single_task`, `parallel_for` etc, but that argument is in our namespace (`ext::oneapi::property_list`); if a developer tries to use this extension with another SYCL compiler, they'll get an error about `ext::oneapi::property_list` not being defined, and from that it will be obvious that they're trying to use an extension.

Arguably we do have the freedom to choose a different function name in this case, but I think the proposed syntax is easier to remember and still in line with the guidelines.

Tagging @gmlueck for awareness, and in case he has anything to add.